### DISCUSSION TITLE
Access the underlying enumerator within a `foreach` statement

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#6836] [Language Ideas]
Currently there is no way to access the underlying enumerator during an enumeration of a collection using `foreach`, which leads to the programmer manually typing out the enumeration process.

An enumerator could expose its current enumeration state that could be useful in some scenarios. There could be a way to allow the user to get the enumerator by defining a variable to hold it within the `foreach` statement itself.

## Example
Assume this enumerator:
```csharp
public class CustomEnumerator<T> : IEnumerator<T>
{
    public int Sequentials { get; private set; }
    public int Invalids { get; private set; }
    // Assume logic is handled in whatever way and these values do something
}
```

Using the enumerator in a `foreach`:
```csharp
// Using the `by` keyword used in LINQ, another keyword could be used
foreach (var item in enumerable by enumerator)
{
    // The GetEnumerator method that allows this enumeration is of a custom type, say CustomEnumerator
    // So automatically, the enumerator that is asssigned to `enumerator` is of CustomEnumerator type

    if (enumerator.Sequentials is 3)
    {
        // invoke an event based on domain-specific logic
        SequentialEvent?.Invoke(item);
    }
}

// Get the Invalids value of the enumerator, which is still in scope
int invalidCount = enumerator.Invalids;
```

## Considerations/Alternatives

Starting with C# 9.0, a custom extension for `IEnumerator` can be defined, allowing direct usage of the given enumerator instance in the `foreach` statement:
```csharp
foreach (var item in enumerator)
// enumerator is IEnumerator<T>, not IEnumerable<T>
```
which enables direct access to the enumerator, as it is locally available for consumption to be used in the statement.

This hacks around the idea of an enumerable, merging the concept of an enumerator and the enumerable itself.

**Comment by svick** [comment]
Enumerators with custom members are incredibly rare. And manually iterating them is less convenient than `foreach`, but not by that much:

```c#
using var enumerator = enumerable.GetEnumerator();
while (enumerator.MoveNext())
{
    if (enumerator.Sequentials is 3)
    {
        // invoke an event based on domain-specific logic
        SequentialEvent?.Invoke(enumerator.Current);
    }
}

// Get the Invalids value of the enumerator, which is still in scope
int invalidCount = enumerator.Invalids;
```

So this syntax sugar doesn't seem worthwhile to me.

**Reply by CyrusNajmabadi** [reply]
I'm in agreement here. The expanded form is very easy and clear to use for this extremely rare case.
